### ISSUE TITLE
Restrict the scope of this project

### ISSUE THREAD

**Issue by julienrf** [scalacenter/scalajs-bundler#314]
I believe we have too many extension points in this plugin (`webpackNodeArgs`, `npmExtraArgs`, `additionalNpmConfig`, `webpackDevServerExtraArgs`, `webpackExtraArgs`, `yarnExtraArgs`). I’m wondering if that’s a sign that the plugin is not at the right “place”.

Maybe we should just generate a webpack configuration file and let users call webpack, npm, yarn, etc. as they want?

Initially, one goal of the plugin was to provide an “integrated” developer experience: no need to bother about writing a `package.json` file, or even maybe a webpack configuration file: just add an `npmDependencies += …` line to your `build.sbt` and you’re done! However, it turns out that this automation is achieved at the expense of (a required) flexibility…

**Comment by julienrf** [comment]
See also: http://appddeevvmeanderings.blogspot.com/2017/11/you-do-not-need-scalajs-bundler-to-use.html

**Comment by ramnivas** [comment]
Could we take this a bit further and re-evaluate if this plugin is needed at all (as discussed in the link the the previous comment)? If the integrated experience is lost, I don't think writing webpack config by hand is that difficult to warrant the existence of this plugin.

**Comment by exoego** [comment]
I agree to this proposition.

For me, scalajs-bundler was very useful for the first-time writing in Scala.js when I had no experimence with webpack things.
Now I learned webpack and created extra `webpack.config.js` for productivity and optimization, so advantage from scalajs-bundler is getting smaller and disadvantange is getting bigger.
One example is that we cannot leverage npm ecosystem like dependabot, noted in #310.

**Comment by julienrf** [comment]
I believe _something_ could still be useful to provide the path of the produced .js file to the (manually written) webpack configuration file.

And also, I think a version of scalajs-bundler with less options, making it super easy to use npm packages, would also make sense, even if we then recommend developers to switch to a manually written webpack configuration file and a manual invocation of webpack.

**Comment by easel** [comment]
I've been thinking along the same lines. Webpack and the js ecosystem prefers to be the top level process, so we're always going to feel pain trying to wrap it from SBT. 

That being said, the js ecosystem relies heavily on easy out of the box experiences -- see create-react-app for example. I'd hesitate to move farther from a good out of the box experience unless we can piggyback on something else.

I think the ideal scenario involves heavily optimizing scala.js for producing libraries and integrating them into the rest of the ecosystem, via a babel plugin for instance. If you could drop in a babel plugin that could handle .scala files in your source tree, then scala-js bundler really just becomes some pull requests to create-react-app implementing a --scalajs option much like typescript has.

**Comment by julienrf** [comment]
@easel But then in case your backend is written in Scala as well (which I guess is often the case if the frontend is in Scala) you have to define another integration step to make the frontend available as a static asset in the server. The nice thing with scalajs-bundler is that it automates that part (with sbt-web).

**Comment by easel** [comment]
@julienrf agreed, which is why I'm still using it. I have it integrated with play, including asset filename mangling and everything just works. It might be helpful to define exactly which use cases it's worth embedding in SBT and tightening scope to those only.

For instance, I don't think any sort of webpack-first experience is going result in a 1-command startup using play.

**Comment by ngbinh** [comment]
One of the thing we need in scala.js bundler is https://scalacenter.github.io/scalajs-bundler/reference.html#bundling-mode-library-and-application . Without it, webpack will choke hard on our tens of MBs js output from scala.js. Is there a way to do so outside of scala.js bundler plugin?

**Comment by easel** [comment]
@ngbinh I agree that a webpack integrated, fastopt development cycle is a key value-add for scalajs-bundler. 

With modern browesrs that support es-modules, it should be possible to output a fastopt es-module and then have webpack inline it on change without attempting to tree-shake it. I've been meaning to expirement with it and see what I can accomplish but haven't gotten around to it yet.

**Comment by nafg** [comment]
What I'm missing from that blog post is how do you manage npm dependencies? I need them to propagate through sbt module dependencies and library dependencies.

**Comment by ngbinh** [comment]
Snowpack looks quite interesting https://www.snowpack.dev/

**Comment by mushtaq** [comment]
I am experimenting with using snowpack with Scala.js. You can try this [example project](https://github.com/mushtaq/scalajs-snowpack-example).

`sbtn` makes it really easy to emulate what `tsc` template does. The example allows to mix Typescript with Scala.js.

I am using Scala.js 1.2.0 which supports `testHtml` for `ESModules`, but it does not work if you `import` other modules (like `rxjs` in the example). I will try and report it.

Another issue, before this becomes usable is `ESModule` support in `scala-js-env-selenium`: https://github.com/scala-js/scala-js-env-selenium/issues/115

If something like this works, it will be a great alternative to webpack/scalajs-bunlder.

cc/ @easel @nafg @ramnivas @julienrf
